### PR TITLE
ASM-3439, ASM-3444 improve idrac firmware update

### PR DESCRIPTION
**Pull request by gavin-scott**
- allow for idrac to go offline during idrac firmware update

- monitor each job individually with a max execution time of 30
  minutes

### CODE REVIEW

**Comment by gavin-scott** [comment]
@asaenz can you review? I'm still running a couple tests on this.

**Comment by gavin-scott** [comment]
Provisionally +1'd from @asaenz looking over my shoulder pending a successful test run.

File: lib/puppet/provider/idrac_fw_installfromuri/wsman.rb [hunk 1]
```diff
@@ -60,11 +60,21 @@ def update(firmware_list)
...
       job_status = 'new'
+      started = Time.now
       until job_status =~ /Downloaded|Completed|Failed/
         sleep 30
-        job_status = checkjobstatus job_id
+        begin
+          job_status = checkjobstatus job_id
+        rescue ASM::WsMan::StandardError => e
+          job_status = 'TemporaryFailure'
+          Puppet.warning("Look up job status for #{job_id} failed: #{e}")
```
Comment (gavin-scott) [inline lib/puppet/provider/idrac_fw_installfromuri/wsman.rb]: @asaenz good call, it was Puppet.warning not Puppet.warn

**Comment by gavin-scott** [comment]
DO NOT MERGE. Needs further testing and review with @logankimmel

**Comment by logankimmel** [comment]
@gavin-scott I had checkout this so I could look over it line by line in the IDE.  There's a lot of tribal knowledge happening here on this firmware update provider but all your changes look good.  I'm gonna test this on a couple components but I don't see anything wrong here

File: lib/puppet/provider/idrac_fw_installfromuri/wsman.rb [hunk 2]
```diff
@@ -49,43 +57,51 @@ def create
...
-    components = []
-    firmware_list.each do |f|
-      components << f["component_id"]
-    end
-    reboot_required = true
-    if components.all? {|c| c.to_s =~ /#{UEFI_DIAGNOSTICS_ID}|#{LC_ID}|#{IDRAC_ID}|#{DRIVER_PACK}/}
+
+    # Reboot if necessary
+    reboot_job_ids = statuses.values.map do |v|
+      v[:job_id] unless NO_REBOOT_COMPONENT_IDS.include?(v[:firmware]['component_id'])
```
Comment (logankimmel) [inline lib/puppet/provider/idrac_fw_installfromuri/wsman.rb]: you'll this should be:
      v[:job_id] unless NO_REBOOT_COMPONENT_IDS.include?(v[:firmware]['component_id'].to_i)
or it won't work correctly

Comment (gavin-scott) [inline lib/puppet/provider/idrac_fw_installfromuri/wsman.rb]: Ah, great catch, thanks! Do you think that could explain why my last run failed to schedule the reboot jobs? The above would have meant that I didn't have any reboot_job_ids when I should have I think. See ASM-3444 for stack trace.

**Comment by gavin-scott** [comment]
Updated with NO_REBOOT_COMPONENT_IDS .to_i fix pointed out by @logankimmel

**Comment by gavin-scott** [comment]
@logankimmel can you do a final review and +1 if it looks ok? I am getting complete success runs now updating most of the components on a server on my appliance at 172.18.9.100

**Comment by logankimmel** [comment]
@gavin-scott +1

**Pull request merged**